src/pipelines/epidemiology/gb_authority.py

```diff
@@ -50,12 +91,46 @@ def parse_dataframes(
 
 
 class Covid19UkL1DataSource(Covid19UkL2DataSource):
-    def parse_dataframes(
-        self, dataframes: Dict[str, DataFrame], aux: Dict[str, DataFrame], **parse_opts
-    ) -> DataFrame:
-        data = super().parse_dataframes(dataframes, aux, **parse_opts)
+    def parse(self, sources: Dict[str, str], aux: Dict[str, DataFrame], **parse_opts) -> DataFrame:
+        data = super().parse(sources, aux, **parse_opts)
 
         # Aggregate data to the country level
         data = data.groupby("date").sum().reset_index()
         data["key"] = "GB"
         return data
+
+
+class Covid19UkL3DataSource(DataSource):
+    """ This API can be used to get historical case data for utla and ltla in the UK """
+
+    def parse(self, sources: Dict[str, str], aux: Dict[str, DataFrame], **parse_opts) -> DataFrame:
+        cases = {
+            "date": "date",
+            "areaCode": "areaCode",
+            "newCasesBySpecimenDate": "newCasesBySpecimenDate",
+            "cumCasesBySpecimenDate": "cumCasesBySpecimenDate",
+        }
+
+        api = Cov19API(filters=["areaType=utla"], structure=cases)
+        utla_json = api.get_json()
+        data = DataFrame.from_dict(utla_json["data"])
+
+        data.areaCode = data.areaCode.apply(
+            lambda x: wales_utla_map[x] if x in wales_utla_map else x
+        )
+        data = data.groupby(["date", "areaCode"], as_index=False).sum()
+
+        data = table_rename(
+            data,
+            {
+                "areaCode": "subregion2_code",
+                "newCasesBySpecimenDate": "new_confirmed",
+                "cumCasesBySpecimenDate": "total_confirmed",
+                "date": "date",
+            },
+            drop=True,
+        )
+
+        data.date = data.date.apply(lambda x: datetime_isoformat(x, "%Y-%m-%d"))
```

Comment:
Beware that sometimes pandas will helpfully convert a column to datetime type automatically. `datetime_isoformat` should convert the values to `str` before parsing the date, but I've seen instances where that didn't work and I have not found the root cause yet.

Reply:
Thanks, it seems that this works fine at the moment, don't see any unmatched data.